### DISCUSSION TITLE
[designspacelib] discrete axis, 2 fonts, how to set AxisValueArrays

### DISCUSSION THREAD

**Discussion by LettError** [fonttools/fonttools#2790] [Q&A]
I'm struggling with a STAT issue and I want to solve it using ds5 and fontmake. It is a typeface with an interpolating `wdth` axis and a discrete `ital` axis. So there are two variable fonts, roman and italic.

The fonts showed some odd behaviour in InDesign. I'm grateful to have received very specific help from Adobe in the form of ttx files for the roman and italic. With these patches applied the fonts work well. 

I would like to incorporate  these values in the designspace file so that I can generate the fonts in one go, without having to patch them afterwards.

The roman (hopefully the relevant bits, the full ttx files are in the screengrab below)
```xml
    <AxisValueArray>
      <AxisValue index="0" Format="3">
        <AxisIndex value="1"/>
        <Flags value="2"/>
        <ValueNameID value="267"/>  <!-- Italic -->
        <Value value="0.0"/>
        <LinkedValue value="1.0"/>
      </AxisValue>
    </AxisValueArray>
    <ElidedFallbackNameID value="2"/>  <!-- Regular -->
```

The italic:
```xml
    <AxisValueArray>
      <AxisValue index="0" Format="3">
        <AxisIndex value="1"/>
        <Flags value="0"/>
        <ValueNameID value="267"/>  <!-- Italic -->
        <Value value="1.0"/>
        <LinkedValue value="0.0"/>
      </AxisValue>
    </AxisValueArray>
    <ElidedFallbackNameID value="267"/>  <!-- Italic -->
```

* I guess the key question is: is it possible to express the values above in the designspace? 
* I can of course provide more data to illustrate this but am hoping this might be enough.

The designspace snippet below does *not* result in the right `AxisValueArray`. I've included it to illustrate where the struggle is.

```xml
  <axes>
    <axis tag="wdth" name="width" minimum="50" maximum="100" default="100">
      <map input="50" output="0"/>
      <map input="55" output="121"/>
      <map input="62" output="255"/>
      <map input="70" output="409"/>
      <map input="79" output="583"/>
      <map input="89" output="779"/>
      <map input="100" output="1000"/>
      <labels>
        <label uservalue="50" userminimum="50" usermaximum="52.99" name="Fifty"/>
        <label uservalue="55" userminimum="53" usermaximum="58.99" name="FiftyFive"/>
        <label uservalue="62" userminimum="59" usermaximum="65.99" name="SixtyTwo"/>
        <label uservalue="70" userminimum="66" usermaximum="74.99" name="Seventy"/>
        <label uservalue="79" userminimum="75" usermaximum="84.99" name="SeventyNine"/>
        <label uservalue="89" userminimum="85" usermaximum="95.99" name="EightyNine"/>
        <label uservalue="100" userminimum="96" usermaximum="100" name="Hundred"/>
      </labels>
    </axis>
    <axis tag="ital" name="italic" values="0 1000" default="0">
      <labels>
        <label uservalue="0" name="Upright" elidable="true"/>
        <label uservalue="1000" name="Italic"/>
      </labels>
    </axis>
  </axes>
```

Maybe TMI, but this is a screengrab of both ttx files.
<img width="1450" alt="image" src="https://user-images.githubusercontent.com/1243961/187385089-0e67fa97-cd46-4562-a05d-7aa21588bddb.png">

Thanks!

Cc @belluzj

**Comment by belluzj** [comment]
Maybe you're just missing the `linkeduservalue="1000"` attribute on your labels? That should turn your AxisValues records into format 3 as per the screenshots.

```xml
    <axis tag="ital" name="italic" values="0 1000" default="0">
      <labels>
        <label uservalue="0" name="Upright" elidable="true" linkeduservalue="1000" />
        <label uservalue="1000" name="Italic" linkeduservalue="0" />
      </labels>
    </axis>
```

I hope this is all there is to it! Documentation:

https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#label-element-axis

![image](https://user-images.githubusercontent.com/3616772/187397043-f34d1e20-3b6c-41f0-ae62-1df5c08381d2.png)

Side note: I was under the impression that you only needed to link the user value in one direction (e.g. upright has linkeduservalue="1000" but not the other way round). Did you get from Adobe that you needed both directions?

**Comment by LettError** [comment]
Thank you!

> Side note: I was under the impression that you only needed to link the user value in one direction (e.g. upright has linkeduservalue="1000" but not the other way round). Did you get from Adobe that you needed both directions?

Hm, I don't understand the issue enough to confirm that. But the behaviour was as follows: in InDesign, text set in Roman, use the slider to change the width, the text would flip to Italic. The first remedy was to add names in `nameID=25`. This fixed the roman-to-italic toggle, but introduced a new italic-to-roman toggle when using the slider. The ttx patches above remedied that.

When I add *both* `linkeduservalues`...
```xml
        <label uservalue="0" name="Upright" elidable="true" linkeduservalue="1000" />
        <label uservalue="1000" name="Italic" linkeduservalue="0" />
```
... fontmake (3.4.0) crashes:
```
(letterror-venv) erik@eBookPro Display % fontmake -m /Users/erik/code/Xxxx.designspace -o variable --flatten-components
Traceback (most recent call last):
  File "/Users/erik/code/ltr_py_env/bin/fontmake", line 8, in <module>
    sys.exit(main())
  File "/Users/erik/code/ltr_py_env/lib/python3.7/site-packages/fontmake/__main__.py", line 639, in main
    project.run_from_designspace(inputs.designspace_path, **args)
  File "/Users/erik/code/ltr_py_env/lib/python3.7/site-packages/fontmake/font_project.py", line 1035, in run_from_designspace
    for discrete_location, subDoc in splitInterpolable(designspace):
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/split.py", line 115, in splitInterpolable
    makeInstanceFilename=makeInstanceFilename,
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/split.py", line 290, in _extractSubSpace
    statNames = getStatNames(doc, instance.getFullUserLocation(doc))
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/statNames.py", line 120, in getStatNames
    regularStatNames = getStatNames(doc, regularUserLocation)
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/statNames.py", line 120, in getStatNames
    regularStatNames = getStatNames(doc, regularUserLocation)
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/statNames.py", line 120, in getStatNames
    regularStatNames = getStatNames(doc, regularUserLocation)
  [Previous line repeated 986 more times]
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/statNames.py", line 67, in getStatNames
    defaultSource: Optional[SourceDescriptor] = doc.findDefault()
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/__init__.py", line 2761, in findDefault
    defaultDesignLocation = self.newDefaultLocation()
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/__init__.py", line 2643, in newDefaultLocation
    axisDescriptor.default
  File "/Users/erik/code/fonttools/Lib/fontTools/designspaceLib/__init__.py", line 931, in map_forward
    from fontTools.varLib.models import piecewiseLinearMap
  File "<frozen importlib._bootstrap>", line 1019, in _handle_fromlist
RecursionError: maximum recursion depth exceeded while calling a Python object
```

When I only include one of `linkeduservalues` either in the roman or the italic, the toggling behaviour returns.

**Comment by punchcutter** [comment]
We define the STAT in the feature file which is what I used to then dump the above ttx. 

> Side note: I was under the impression that you only needed to link the user value in one direction (e.g. upright has linkeduservalue="1000" but not the other way round). Did you get from Adobe that you needed both directions?

No, nobody said both directions are required and they aren't. This was just one quick way to make this particular set of fonts work in InDesign, but there are other possible ways. Like adding Format 2 range values to both fonts and giving them unique names. The reason moving the slider can make it jump from Italic to Regular is that both fonts contained the exact same STAT table with no differentiation and no AxisValue records. InDesign is using the STAT information to construct location names on the fly as you move the slider. In this case with identical STAT tables it can't know which one you mean so it switches to the non-Italic (or more accurately the first font). With the `ital` linking set up as well as correct `ElidedFallbackNameID` (as in different names for each font because there are no other AxisValue names to use when constructing a name here...) there shouldn't be problems with InDesign.

**Comment by belluzj** [comment]
Thanks for the longer explanation @punchcutter .

I think the recursion error in fontmake is due to me making the assumption that there would be a linked value in only one direction, and having them both creates a loop of following the linked values... I should open a PR to prevent that from happening.

**Comment by jenskutilek** [comment]
Tangential:

```xml
      <labels>
        <label uservalue="50" userminimum="50" usermaximum="52.99" name="Fifty"/>
        <label uservalue="55" userminimum="53" usermaximum="58.99" name="FiftyFive"/>
        <label uservalue="62" userminimum="59" usermaximum="65.99" name="SixtyTwo"/>
        <label uservalue="70" userminimum="66" usermaximum="74.99" name="Seventy"/>
        <label uservalue="79" userminimum="75" usermaximum="84.99" name="SeventyNine"/>
        <label uservalue="89" userminimum="85" usermaximum="95.99" name="EightyNine"/>
        <label uservalue="100" userminimum="96" usermaximum="100" name="Hundred"/>
      </labels>
```

@LettError I think you can set the usermaximum to the same value as the userminimum of the next range. The spec explicitly states that in that case, at the value the higher range takes precedence. The way you set it now, there may be a tiny gap between the ranges in the final values.

**Reply by LettError** [reply]
Thank you, I will try that. This came about when there were a bunch of other things not working and I was trying many things.
